### PR TITLE
MG-887 - Add Users CLI Tests.

### LINKED ISSUE #887: Create CLI tests
Only Users service is tested, we also need CLI tests for other services.

### LINKED ISSUE #2253: Create Users CLI tests
### Is your feature request related to a problem? Please describe.

There are currently missing tests for CLI, one of the tests required is tests for users in CLI.

### Describe the feature you are requesting, as well as the possible use case(s) for it.

Add tests for users on CLI. The tests should cover inputs passed to the CLI and check responses logged.

### Indicate the importance of this feature to you.

Should-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by WashingtonKK**
# What type of PR is this?

This is an enhancement because it adds cli tests for users.

## What does this do?

This adds tests for cli, to check for invalid args input, and logging of output.

## Which issue(s) does this PR fix/relate to?

- Related to #887
- Resolves #2253

## Have you included tests for your changes?

Yes, I have included tests for my changes.

## Did you document any new/modified feature?

No, I have not updated the documentation because there are no new features, just an increase in scope of the existing tests to cover CLI.

### Notes

This is the first PR in this series, to be merged first.

### CODE REVIEW

File: cli/users_test.go [hunk 1]
```diff
@@ -0,0 +1,926 @@
+// Copyright (c) Abstract Machines
+// SPDX-License-Identifier: Apache-2.0
+
+package cli
```
Comment (rodneyosodo) [inline cli/users_test.go]: This should be `cli_test`

File: cli/utils.go [hunk 2]
```diff
@@ -97,3 +99,134 @@ func convertMetadata(m string) (map[string]interface{}, error) {
 	}
 	return nil, nil
 }
+
+func executeCommand(root *cobra.Command, v any, args ...string) (errorRes, out string, err error) {
```
Comment (rodneyosodo) [inline cli/utils.go]: This should not be in cli package but tests package

File: cli/utils.go [hunk 3]
```diff
@@ -97,3 +99,134 @@ func convertMetadata(m string) (map[string]interface{}, error) {
...
+	buf := new(bytes.Buffer)
+	root.SetOut(buf)
+	root.SetErr(buf)
+	root.SetArgs(args)
+
+	o := os.Stdout
```
Comment (rodneyosodo) [inline cli/utils.go]: Are we using these variables?

File: cli/utils.go [hunk 4]
```diff
@@ -97,3 +99,134 @@ func convertMetadata(m string) (map[string]interface{}, error) {
...
+	e := os.Stderr
+
+	r, w, _ := os.Pipe()
```
Comment (rodneyosodo) [inline cli/utils.go]: Don't ignore errors assert they are nil or return

File: cli/utils.go [hunk 5]
```diff
@@ -97,3 +99,134 @@ func convertMetadata(m string) (map[string]interface{}, error) {
...
+	}
+
+	res := outBuf.Bytes()
+
+	_, err = errBuf.ReadFrom(r1)
+	if err != nil {
+		return "", buf.String(), err
+	}
+
+	err = json.Unmarshal(res, v)
```
Comment (rodneyosodo) [inline cli/utils.go]: inline

File: cli/users_test.go [hunk 6]
```diff
@@ -0,0 +1,926 @@
...
+	}
+
+	for _, tc := range cases {
+		var pg mgsdk.GroupsPage
+		sdkCall := sdkmocks.On("ListUserGroups", mock.Anything, mock.Anything, mock.Anything).Return(tc.page, tc.sdkerr)
+		errRes, out, err := executeCommand(rootCmd, &pg, tc.args...)
+
+		assert.Nil(t, err, fmt.Sprintf("unexpected error when running command: %s, got error: %v, with args: %v", tc.desc, err, tc.args))
+		assert.Equal(t, tc.error, errRes, fmt.Sprintf("%s unexpected error response: expected %s got error: %s", tc.desc, tc.error, errRes))
+		assert.True(t, !strings.Contains(out, rootCmd.Use), fmt.Sprintf("%s invalid usage: %s", tc.desc, out))
```
Comment (rodneyosodo) [inline cli/users_test.go]: assert False

File: cli/users_test.go [hunk 7]
```diff
@@ -0,0 +1,926 @@
...
+				domainsCommand,
+				user.ID,
+				validToken,
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		var pg mgsdk.GroupsPage
+		sdkCall := sdkmocks.On("ListUserGroups", mock.Anything, mock.Anything, mock.Anything).Return(tc.page, tc.sdkerr)
```
Comment (rodneyosodo) [inline cli/users_test.go]: Reduce use of `mock.Anything`

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: cli/utils.go (hunk 3 above)
Comment (WashingtonKK) [reply cli/utils.go]: Yes, `o` and `e` store the initial outputs and they are restored to the original value after use.

**Review by rodneyosodo — CHANGES_REQUESTED** [review]
Update `make test` to include this tests https://github.com/absmach/magistrala/blob/42be65a3d20df7ad999253e5b2b2a10490b1d64d/Makefile#L127 and also trigger it to run on PRs https://github.com/absmach/magistrala/blob/main/.github/workflows/tests.yml

**Review by arvindh123 — CHANGES_REQUESTED** [review]
At logs, we directly printing to `os.stdout` and `os.stderr`, 
Actually, cobra should take care to redirecting to out or error 

So We need to modify the log to like below 

```go
func logError( cmd *cobra.Command , err error) {
	boldRed := color.New(color.FgRed, color.Bold)
	boldRed.Fprintf(cmd.OutOrStderr(, "\nerror: ")

	fmt.Fprintf(cmd.OutOrStderr(), "%s\n\n", color.RedString(err.Error()))
}
```
Please modify the log functions like above and test

if it works , then create a `custom cobra.Command`, like below

```go
type cmd struct {
	*cobra.Command
}

func (c *cmd) logJSON(iList ...interface{}) {
	for _, i := range iList {
		m, err := json.Marshal(i)
		if err != nil {
			c.logError(err)
			return
		}

		pj, err := prettyjson.Format(m)
		if err != nil {
			c.logError(err)
			return
		}

		fmt.Fprintf(c.OutOrStdout(), "\n%s\n\n", string(pj))
	}
}

func (c *cmd) logUsage(u string) {
	fmt.Fprintf(c.OutOrStdout(), color.YellowString("\nusage: %s\n\n"), u)
}

func (c *cmd) logError(err error) {
	boldRed := color.New(color.FgRed, color.Bold)
	boldRed.Fprintf(c.OutOrStderr(), "\nerror: ")

	fmt.Fprintf(c.OutOrStderr(), "%s\n\n", color.RedString(err.Error()))
}
```

**Comment by WashingtonKK** [comment]
> At logs, we directly printing to `os.stdout` and `os.stderr`, Actually, cobra should take care to redirecting to out or error
> 
> So We need to modify the log to like below
> 
> ```go
> func logError( cmd *cobra.Command , err error) {
> 	boldRed := color.New(color.FgRed, color.Bold)
> 	boldRed.Fprintf(cmd.OutOrStderr(, "\nerror: ")
> 
> 	fmt.Fprintf(cmd.OutOrStderr(), "%s\n\n", color.RedString(err.Error()))
> }
> ```
> 
> Please modify the log functions like above and test
> 
> if it works , then create a `custom cobra.Command`, like below
> 
> ```go
> type cmd struct {
> 	*cobra.Command
> }
> 
> func (c *cmd) logJSON(iList ...interface{}) {
> 	for _, i := range iList {
> 		m, err := json.Marshal(i)
> 		if err != nil {
> 			c.logError(err)
> 			return
> 		}
> 
> 		pj, err := prettyjson.Format(m)
> 		if err != nil {
> 			c.logError(err)
> 			return
> 		}
> 
> 		fmt.Fprintf(c.OutOrStdout(), "\n%s\n\n", string(pj))
> 	}
> }
> 
> func (c *cmd) logUsage(u string) {
> 	fmt.Fprintf(c.OutOrStdout(), color.YellowString("\nusage: %s\n\n"), u)
> }
> 
> func (c *cmd) logError(err error) {
> 	boldRed := color.New(color.FgRed, color.Bold)
> 	boldRed.Fprintf(c.OutOrStderr(), "\nerror: ")
> 
> 	fmt.Fprintf(c.OutOrStderr(), "%s\n\n", color.RedString(err.Error()))
> }
> ```

@arvindh123 , this implementation does not work, and coverage is still not generated.

**Comment by WashingtonKK** [comment]
> > At logs, we directly printing to `os.stdout` and `os.stderr`, Actually, cobra should take care to redirecting to out or error
> > So We need to modify the log to like below
> > ```go
> > func logError( cmd *cobra.Command , err error) {
> > 	boldRed := color.New(color.FgRed, color.Bold)
> > 	boldRed.Fprintf(cmd.OutOrStderr(, "\nerror: ")
> > 
> > 	fmt.Fprintf(cmd.OutOrStderr(), "%s\n\n", color.RedString(err.Error()))
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Please modify the log functions like above and test
> > if it works , then create a `custom cobra.Command`, like below
> > ```go
> > type cmd struct {
> > 	*cobra.Command
> > }
> > 
> > func (c *cmd) logJSON(iList ...interface{}) {
> > 	for _, i := range iList {
> > 		m, err := json.Marshal(i)
> > 		if err != nil {
> > 			c.logError(err)
> > 			return
> > 		}
> > 
> > 		pj, err := prettyjson.Format(m)
> > 		if err != nil {
> > 			c.logError(err)
> > 			return
> > 		}
> > 
> > 		fmt.Fprintf(c.OutOrStdout(), "\n%s\n\n", string(pj))
> > 	}
> > }
> > 
> > func (c *cmd) logUsage(u string) {
> > 	fmt.Fprintf(c.OutOrStdout(), color.YellowString("\nusage: %s\n\n"), u)
> > }
> > 
> > func (c *cmd) logError(err error) {
> > 	boldRed := color.New(color.FgRed, color.Bold)
> > 	boldRed.Fprintf(c.OutOrStderr(), "\nerror: ")
> > 
> > 	fmt.Fprintf(c.OutOrStderr(), "%s\n\n", color.RedString(err.Error()))
> > }
> > ```
> 
> @arvindh123 , this implementation does not work, and coverage is still not generated.

I suppose this is because the function `OutOrStdout` simply returns the writer that is set using `SetOut` and `SerErr` and uses it to write all output which still does not solve our issue.

File: cli/setup_test.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+		res := outputBuffer.Bytes()
+		assert.Greater(t, len(res), 0, "Error reading from pipe")
+		err = json.Unmarshal(res, entity)
+		assert.NoError(t, err, "Error unmarshalling entity")
+	default:
+		return ""
+	}
+
+	return ""
+}
```
Comment (arvindh123) [inline cli/setup_test.go:69]: Some thing like this works ,But we need to mock the actual service server,
```suggestion
func executeCommand(t *testing.T, root *cobra.Command, args ...string) string {
	buffer := new(bytes.Buffer)
	root.SetOut(buffer)
	root.SetErr(buffer)
	root.SetArgs(args)

	err := root.Execute()
	assert.NoError(t, err, "Error executing command")

	return buffer.String()
}
```

File: cli/users_test.go [hunk 9]
```diff
@@ -0,0 +1,1124 @@
...
+			desc: "create user successfully without token",
+			args: []string{
+				createCommand,
+				"john doe",
+				"john.doe@example.com",
+				"12345678",
+			},
+			user:    user,
+			logType: entityLog,
+		},
```
Comment (arvindh123) [inline cli/users_test.go:69]: Wrong args are passed.
```suggestion
			desc: "create user successfully with token",
			args: []string{
				createCommand,
				user.Name,
				user.Credentials.Identity,
				user.Credentials.Secret,
				token,
			},
			user:    user,
			logType: entityLog,
		},
```

File: cli/users_test.go [hunk 10]
```diff
@@ -0,0 +1,1124 @@
...
+		{
+			desc: "create user successfully without token",
+			args: []string{
+				createCommand,
+				"john doe",
+				"john.doe@example.com",
+				"12345678",
+			},
```
Comment (arvindh123) [inline cli/users_test.go:66]: Same here wrong args 
```suggestion
				createCommand,
				user.Name,
				user.Credentials.Identity,
				user.Credentials.Secret,
			},
```

File: cli/users_test.go [hunk 11]
```diff
@@ -0,0 +1,1124 @@
...
+			logType: entityLog,
+		},
+		{
+			desc: "failed to create user",
+			args: []string{
+				createCommand,
+				"john doe",
+				"john.doe@example.com",
+				"12345678",
+			},
```
Comment (arvindh123) [inline cli/users_test.go:77]:
```suggestion
				user.Name,
				user.Credentials.Identity,
				user.Credentials.Secret,
				token,
```

File: cli/users_test.go [hunk 12]
```diff
@@ -0,0 +1,1124 @@
...
+			sdkerr:        errors.NewSDKErrorWithStatus(svcerr.ErrCreateEntity, http.StatusUnprocessableEntity),
+			errLogMessage: fmt.Sprintf("\nerror: %s\n\n", errors.NewSDKErrorWithStatus(svcerr.ErrCreateEntity, http.StatusUnprocessableEntity).Error()),
+			logType:       errLog,
+		},
+		{
+			desc:    "create user with invalid args",
+			args:    []string{createCommand, user.Name, user.Credentials.Identity},
+			logType: usageLog,
```
Comment (arvindh123) [inline cli/users_test.go:85]:
```suggestion
			desc: "create user with invalid args",
			args: []string{
				createCommand,
				user.Name,
				user.Credentials.Identity,
			},
			logType: usageLog,
```

File: cli/users_test.go [hunk 13]
```diff
@@ -0,0 +1,1124 @@
...
+				Credentials: mgsdk.Credentials{
+					Identity: tc.args[1],
+					Secret:   tc.args[2],
+				},
+			}
+			sdkCall = sdkMock.On("CreateUser", mock.Anything, sdkUser).Return(tc.user, tc.sdkerr)
+		}
+		var usr mgsdk.User
+		out := executeCommand(t, rootCmd, &usr, tc.logType, tc.args...)
+
```
Comment (arvindh123) [inline cli/users_test.go:103]:
```suggestion
		out := executeCommand(t, rootCmd, tc.args...)

```

**Review by arvindh123 — CHANGES_REQUESTED** [review]
This one works for me, We need to some how mock cobra execute or we need to create mock server service which cli is calling
```go
func logJSONCmd(cmd cobra.Command, iList ...interface{}) {
	for _, i := range iList {
		m, err := json.Marshal(i)
		if err != nil {
			logErrorCmd(cmd, err)
			return
		}

		pj, err := prettyjson.Format(m)
		if err != nil {
			logErrorCmd(cmd, err)
			return
		}

		fmt.Fprintf(cmd.OutOrStdout(), "\n%s\n\n", string(pj))
	}
}

func logUsageCmd(cmd cobra.Command, u string) {
	fmt.Fprintf(cmd.OutOrStdout(), color.YellowString("\nusage: %s\n\n"), u)
}

func logErrorCmd(cmd cobra.Command, err error) {
	boldRed := color.New(color.FgRed, color.Bold)
	boldRed.Fprintf(cmd.ErrOrStderr(), "\nerror: ")

	fmt.Fprintf(cmd.ErrOrStderr(), "%s\n\n", color.RedString(err.Error()))
}

func logOKCmd(cmd cobra.Command) {
	fmt.Fprintf(cmd.OutOrStdout(), "\n%s\n\n", color.BlueString("ok"))
}

func logCreatedCmd(cmd cobra.Command, e string) {
	if RawOutput {
		fmt.Fprintln(cmd.OutOrStdout(), e)
	} else {
		fmt.Fprintf(cmd.OutOrStdout(), color.BlueString("\ncreated: %s\n\n"), e)
	}
}

func logRevokedTimeCmd(cmd cobra.Command, t time.Time) {
	if RawOutput {
		fmt.Fprintln(cmd.OutOrStdout(), t)
	} else {
		fmt.Fprintf(cmd.OutOrStdout(), color.BlueString("\nrevoked: %v\n\n"), t)
	}
}


```

**Comment by dborovcanin** [comment]
Closing as duplicate of https://github.com/absmach/magistrala/pull/2319.

**Pull request closed without merging**